### ISSUE TITLE
Python binding with Eigen / numpy mapping

### ISSUE THREAD

**Issue by brianzhang01** [OxfordRSE/ASMC#17]
This could be nice to avoid a memory copy when trying to use the Python binding return values.

**Comment by brianzhang01** [comment]
I ran the Python line_profiler on my code and here's the timings for the relevant segment.

```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   151                                               @profile
   152                                               def decode_pair(i, j):
   153     44850     155775.0      3.5      0.7          pair_obs = asmc.makePairObs(data.individuals[i // 2], i % 2 + 1,
   154     44850     171810.0      3.8      0.8                                      data.individuals[j // 2], j % 2 + 1)
   155                                                   # a = decoding_quantities.expectedTimes[0] # use this to capture decoding_quantities
   156     44850   15764393.0    351.5     72.6          decode_result = hmm.decodeSummarize(pair_obs)
   157     44850    5609242.0    125.1     25.8          return site_positions, np.array(decode_result[0]), np.array(decode_result[1])
```

`hmm.decodeSummarize()` computes the posterior, summarizes it into a MAP (maximum a posteriori) vector and posterior mean vector, then returns those via the pybind11 interface. This takes 73% of the time. 26% of the time is then spent converting these values to numpy.

**Comment by abhidg** [comment]
@brianzhang01 I'll be starting work on this issue. I think porting `sumOverPairs` to `Eigen::Array` would do the trick. Pybind has mappings to represent Eigen matrices as numpy arrays without copying.

**Comment by brianzhang01** [comment]
Ok, yes I think this is an important part of a successful Python binding. A few thoughts as we go into this.
- I think switching over DecodingReturnValues::sumOverPairs, sumOverPairs00, sumOverPairs01, sumOverPairs11 would be a great first step.
- In my workflow, I currently use HMM::decodeSummarize, a function I wrote myself, which would not be covered by this change. However, I'm also likely to change HMM::decodeSummarize, so I'll probably just do the Eigen / numpy thing myself once I see what you've done (the work you put in would be helpful as an example).
- An ambitious goal is to get rid of all the std::vector related types that are introduced using py::bind_vector in the API. We can keep around VectorIndividual and VectorPairObservations which don't make sense in numpy, but in theory we can get rid of
```
    py::bind_vector<std::vector<bool>>(m, "VectorBool");
    py::bind_vector<std::vector<float>>(m, "VectorFloat");
    py::bind_vector<std::vector<uint>>(m, "VectorUInt");
    py::bind_vector<std::vector<std::vector<float>>>(m, "Matrix");
```
- If we did this, there's a lot more to get moved to Eigen / numpy, i.e. everything in DecodingQuantities. Most of these attributes arguably shouldn't be public, they're more likely friend of HMM, but it can be useful to look up DecodingQuantities::expectedTimes and discretization from Python.
- Starting to move things to Eigen / numpy is a pretty big change. We have one regression test but should be conscious that the process might introduce some bugs that we'll need to track down.
- We'll also want to think through how Eigen will be included with CMake, but I'm guessing this is straightforward

As a next step, I think sumOverPairs would be great, as that's how most users of the Python API will want to call ASMC. After that I may want to go through the Python API with Pier and come up with a list of what we want publicly exposed through Python and what we don't.
